lib/cn/csched_sp3_work.c

```diff
@@ -173,33 +175,48 @@ sp3_work_wtype_root(
     runlen = 1;
 
     /* Look for a contiguous sequence of non-busy kvsets.
-     *
-     * TODO: Starting with the first non-busy kvset, count the number of
-     * kvsets contiguous from the first that would all spill to the same
-     * leaf node.
      */
+
+    n = cn_kvset_can_zspill(le->le_kvset, rmap);
+    if (n)
+        znode = n;
+
     while ((le = list_prev_entry_or_null(le, le_link, &tn->tn_kvset_list))) {
         if (kvset_get_workid(le->le_kvset) != 0)
             break;
 
+        if (znode) {
+            n = cn_kvset_can_zspill(le->le_kvset, rmap);
+            if (n) {
+                if (n->tn_nodeid == znode->tn_nodeid && runlen < runlen_max) {
+                    ++runlen;
+                    continue;
+                } else {
+                    assert(runlen && runlen <= runlen_max);
+                    *rule = CN_RULE_ZSPILL;
+                    return runlen;
+                }
+            }
+        }
```

Comment:
You can eliminate of the runlen_max check for zspill.

Reply:
Hmm, a zspill may have to fall back on an rspill. In which case a long list of kvsets would be a problem. So I added the runlen_max check.